addons/mail_plugin/controllers/mail_plugin.py

```diff
@@ -0,0 +1,290 @@
+            return {'id': -1}
+
+        return {
+                    'id': company.id,
+                    'name': company.name,
+                    'phone': company.phone,
+                    'mobile': company.mobile,
+                    'email': company.email,
+                    'image': company.image_1920,
+                    'address': {
+                        'street': company.street,
+                        'city': company.city,
+                        'zip': company.zip,
+                        'country': company.country_id.name if company.country_id else ''
+                    },
+                    'website': company.website,
+                    'additionalInfo': json.loads(company.iap_enrich_info) if company.iap_enrich_info else {}
+                }
+
+    def _create_company_from_iap(self, domain):
+        iap_data = self._iap_enrich(domain)
+        if 'enrichment_info' in iap_data:
+            return None, iap_data['enrichment_info']
+
+        phone_numbers = iap_data.get('phone_numbers')
+        emails = iap_data.get('email')
+        new_company_info = {
+            'is_company': True,
+            'name': iap_data.get("name") or domain,
+            'street': iap_data.get("street_name"),
+            'city': iap_data.get("city"),
+            'zip': iap_data.get("postal_code"),
+            'phone': phone_numbers[0] if phone_numbers else None,
+            'website': iap_data.get("domain"),
+            'email': emails[0] if emails else None
+        }
+
+        logo_url = iap_data.get('logo')
+        if logo_url:
+            try:
+                response = requests.get(logo_url, timeout=2)
+                if response.ok:
+                    new_company_info['image_1920'] = base64.b64encode(response.content)
+            except Exception as e:
+                _logger.warning('Download of image for new company %s failed, error %s', new_company_info.name, e)
+
+        if iap_data.get('country_code'):
+            country = request.env['res.country'].search([('code', '=', iap_data['country_code'].upper())])
+            if country:
+                new_company_info['country_id'] = country.id
+                if iap_data.get('state_code'):
+                    state = request.env['res.country.state'].search([
+                        ('code', '=', iap_data['state_code']),
+                        ('country_id', '=', country.id)
+                    ])
+                    if state:
+                        new_company_info['state_id'] = state.id
+
+        new_company_info['iap_enrich_info'] = json.dumps(iap_data)
+        new_company = request.env['res.partner'].create(new_company_info)
+        new_company.message_post_with_view(
+            'iap_mail.enrich_company',
+            values=iap_data,
+            subtype_id=request.env.ref('mail.mt_note').id,
+        )
+
+        return new_company, {'type': 'company_created'}
+
+    def _prepare_partner_values(self, partner):
+        return {
```

Comment:
Could probably be simplified with a loop-based dict creation (as hinted above for companies).

Reply:
Done.